db/clientlist_virtual_reader.cc

```diff
@@ -0,0 +1,237 @@
+        const auto rhs_bytes = data_type_for<port_type>()->decompose(rhs);
+        return operator()(lhs, clustering_key_prefix::from_single_value(*_schema, rhs_bytes));
+    }
+};
+}  // anon. namespace
+
+static future<std::vector<system_keyspace::client_data>> get_all_clients_data(
+        distributed<cql_transport::cql_server>& distributed_cql_server) {
+    return distributed_cql_server.map_reduce(clients_reducer{},
+            [] (const cql_transport::cql_server& srv) {
+        const auto& conn_list = srv.get_connections_list();
+        return do_with(std::vector<system_keyspace::client_data>{},
+                [&conn_list] (auto& per_shard_result) {
+            return do_for_each(conn_list.begin(), conn_list.end(),
+                    [&per_shard_result] (const auto& conn) {
+                system_keyspace::client_data cd;
+                cd.ip = conn.get_client_state().get_client_address().addr();
+                cd.port = conn.get_client_state().get_client_port();
+                cd.connection_stage = conn.get_client_state().get_auth_state();
+                cd.protocol_version = conn.get_protocol_version();
+                cd.username = conn.get_client_state().user()->name;
+
+                per_shard_result.emplace_back(std::move(cd));
+            }).then([psr = std::move(per_shard_result)] {
+                return make_ready_future<std::vector<system_keyspace::client_data>>(std::move(psr));
+            });
+        });
+    });
+}
+
+// Clusters clients by IP and sorts them by port number within clusters.
+future<std::vector<system_keyspace::client_data>>
+clientlist_mutation_reader::filter_and_sort_by_port(std::vector<system_keyspace::client_data>&& clients) const {
+    std::vector<system_keyspace::client_data> sorted_clients;
+    sorted_clients.reserve(clients.size());
+    
+    std::unordered_map<
+            decltype(system_keyspace::client_data::ip),
+            std::vector<system_keyspace::client_data>> pk_clusters;
+    
+    for (auto&& client : clients) {
+        const auto pk = partition_key::from_single_value(*_schema, inet_addr_type->decompose(client.ip));
+        const dht::token calculated_token = dht::global_partitioner().get_token(*_schema, pk);
+        if (dht::shard_of(calculated_token) != engine().cpu_id()) {
+            continue;
+        }
+        for (auto& range : _slice.row_ranges(*_schema, pk)) {
+            const auto ck_bytes = data_type_for<decltype(system_keyspace::client_data::port
+                    )>()->decompose(client.port);
+            const auto ck_prefix = clustering_key_prefix::from_single_value(*_schema, ck_bytes);
+
+            if (range.contains(ck_prefix, ck_comparator(_schema))) {
+                pk_clusters[client.ip].emplace_back(std::move(client));
+                break;
+            }
+        }
+    }
+
+    for (auto& [ip, cluster] : pk_clusters) {
+        if (_slice.options.contains(query::partition_slice::option::reversed)) {
```

Comment:
Implementing a reader is not easy, since the interface puts lots of requirements on the implementation. Why not rely on existing implementation? You could put all the mutations into a temporary `memtable` and return `memtable::make_flat_reader()`.

Reply:
But isn't "all the mutations" equivalent to "all the connections"? This could be larger than a page, so we'd be doing O(nr connections) work instead of O(page's worth of connections) work.